account/lnurl.go

```diff
@@ -200,3 +222,217 @@ func (a *Service) getLNURLAuthKey() (*bip32.Key, error) {
+
+	if invoice.DescriptionHash == nil {
+		return nil, errors.New("Description hash not found in invoice.")
+	}
+
+	// FIXME Calculate this hash correctly.
+	// 1. Make a hash as follows: `sha256(utf8ByteArray(unescaped_metadata_string))`.
+	/*
+	   if sha256.Sum256([]byte(fmt.Sprintf("%s", params.Metadata))) != *invoice.DescriptionHash {
+	       return nil, errors.New("Invoice description hash does not match metadata.")
+	   }
+	*/
+
+	/* 8. `LN WALLET` Verifies that amount in provided invoice equals an amount previously specified by user.
+	   amount from client is in millisats.
+	*/
+
+	a.log.Info("lnurl-pay: verify invoice.amount == params.Amount.")
+	if params.Amount != uint64(*invoice.MilliSat) {
+		return nil, errors.New("Invoice amount does not match the amount set by user.")
+	}
+
+	// TODO 9. If routes array is not empty: verifies signature for every provided `ChannelUpdate`, may use these routes if fee levels are acceptable.
+	// ref. https://github.com/lightningnetwork/lightning-rfc/blob/master/07-routing-gossip.md#the-channel_update-message
+	// ref. github.com\lightningnetwork\lnd\lnwire\channel_update.go
+	// FINDOUT Are these routes different from the ones in the invoice? If yes, why?
+	// FINDOUT If these routes can be different, how do I use them in the client?
+	/*
+	   if len(payResponse2.Routes) > 0 {
+	       for _, route := range payResponse2.Routes {
+	           for _, hop := range r {
+
+	               // ref. github.com/lightningnetwork/lnd/lnwire/channel_update.go
+	               // ref. github.com/lightningnetwork/lnd/lnwire/message.go
+	               // hop.ChannelUpdate string
+	               // lnwire.ReadMessage validates a Lightning message.
+	               _, err := lnwire.ReadMessage(strings.NewReader(hop.ChannelUpdate), 0) // TODO What should version number be?
+	               if err != nil {
+	                   return nil, fmt.Errorf("Unable to parse ChannelUpdate message for NodeId %s.", h.NodeId)
+	               }
+
+	           }
+	       }
+	   }
+	*/
+
+	// 10. If `successAction` is not null: `LN WALLET` makes sure that `tag` value of is of supported type, aborts a payment otherwise.
+	sa := payResponse2.SuccessAction
+        var _sa *data.SuccessAction
+	if sa != nil {
+
+		if t := sa.Tag; t != "message" &&
+			t != "url" &&
+			t != "aes" {
+			return nil, fmt.Errorf("Unknown SuccessAction: %s", t)
+		}
+
+                _sa = &data.SuccessAction{
+			Tag:         sa.Tag,
+			Description: sa.Description,
+			URL:         sa.URL,
+			Message:     sa.Message,
+			Ciphertext:  sa.Ciphertext,
+			IV:          sa.IV,
+		}
+
+	}
+
+	// 11. `LN WALLET` pays the invoice, no additional user confirmation is required at this point.
+	// a.lnurlSuccessAction = sa // FIXME Handle this in client.
+	return &data.LNUrlPayInfo{
+		Invoice: payResponse2.PR,
+		SuccessAction: _sa,
+	},
+	nil
+
+}
+
+func (a *Service) decryptLNUrlPayMessage(preImage []byte) (m string, err error) {
```

Comment:
Is this in use?

Reply:
> Is this in use?

Yes. This has all the data + invoice that should be displayed to the user before payment confirmation.

Reply:
> Is this in use?

Not yet. It will be used for SuccessAction.